python/pyspark/sql/dataframe.py

```diff
@@ -1648,8 +1650,30 @@ def toPandas(self):
         0    2  Alice
         1    5    Bob
         """
-        import pandas as pd
-        return pd.DataFrame.from_records(self.collect(), columns=self.columns)
+        if self.sql_ctx.getConf("spark.sql.execution.arrow.enable", "false").lower() == "true":
+            try:
+                import pyarrow
+                tables = self._collectAsArrow()
+                table = pyarrow.concat_tables(tables)
+                return table.to_pandas()
```

Comment:
`toPandas()` is a collect operation on the driver, so all data will have to be present before the pandas.DataFrame can be made regardless.  As @leifwalsh pointed out, the batch size limiting is to benefit the JVM running the executor, which might have different limitations that the driver.